File: clients/javascript/src/resources/blocks.ts
import type { HieroProxyClient } from "../client";
import type { BlockResponse } from "./blocks.types";
import { requireId } from "./validation";

/**
 * Blocks resource — query blocks by number or hash.
 */
export class BlocksResource {
  constructor(private readonly client: HieroProxyClient) {}

  /**
   * List blocks (most recent first).
   */
  async list(): Promise<BlockResponse[]> {
    return this.client.get("/api/v1/blocks");
  }

  /**
   * Get a block by its sequential number.
   */
  async getByNumber(blockNumber: number): Promise<BlockResponse> {
    if (blockNumber < 0) {
      throw new Error("blockNumber must not be negative");
    }
    return this.client.get(`/api/v1/blocks/number/${blockNumber}`);
  }

  /**
   * Get a block by its EVM-compatible hash.
   */
  async getByHash(hash: string): Promise<BlockResponse> {
    const h = requireId(hash, "hash");
    return this.client.get(`/api/v1/blocks/hash/${h}`);
  }
}

File: clients/python/src/hiero_proxy/resources/blocks.py
"""Blocks resource."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from hiero_proxy.resources.validation import require_id

if TYPE_CHECKING:
    from hiero_proxy.client import AsyncHieroProxyClient, HieroProxyClient


class BlocksResource:
    """Block queries — list, get by number or hash."""

    def __init__(self, client: HieroProxyClient | AsyncHieroProxyClient) -> None:
        self._client = client

    def list(self) -> Any:
        return self._client.get("/api/v1/blocks")

    def get_by_number(self, block_number: int) -> Any:
        if block_number < 0:
            raise ValueError("block_number must not be negative")
        return self._client.get(f"/api/v1/blocks/number/{block_number}")

    def get_by_hash(self, block_hash: str) -> Any:
        h = require_id(block_hash, "block_hash")
        return self._client.get(f"/api/v1/blocks/hash/{h}")
